code/modules/food_and_drinks/kitchen_machinery/kitchen_machine.dm

```diff
@@ -361,97 +318,190 @@
+		ffuu.reagents.add_reagent("carbon", amount/2)
+		ffuu.reagents.add_reagent("????", amount/15)
 		ffuu.forceMove(get_turf(src))
 
-/obj/machinery/kitchen_machine/Topic(href, href_list)
-	if(..() || panel_open)
+/obj/machinery/kitchen_machine/update_icon_state()
+	. = ..()
+	if(dirty == 100)
+		icon_state = dirty_icon
 		return
-
-	usr.set_machine(src)
 	if(operating)
-		updateUsrDialog()
-		return
-
-	switch(href_list["action"])
-		if("cook")
-			cook()
-
-		if("dispose")
-			dispose()
-	return
+		icon_state = on_icon
+	else
+		icon_state = off_icon
 
 /obj/machinery/kitchen_machine/build_reagent_description(mob/user)
 	return
 
-/obj/machinery/kitchen_machine/examine(mob/user, infix = "", suffix = "")
+/obj/machinery/kitchen_machine/examine(mob/user)
 	. = ..()
 	var/dat = format_content_descs()
 	if(dat)
 		. += "It contains: <BR>[dat]"
+
+/obj/machinery/kitchen_machine/attack_hand(mob/user)
+	if(stat & (BROKEN|NOPOWER) || panel_open || !anchored)
+		return
+
+	ui_interact(user)
+
+/obj/machinery/kitchen_machine/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = TRUE, datum/tgui/master_ui = null, datum/ui_state/state = GLOB.default_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "KitchenMachine",  name, 400, 300, master_ui, state)
+		ui.open()
+
+/obj/machinery/kitchen_machine/ui_data(mob/user)
+	var/list/data = list()
+	data["name"] = name
+	data["operating"] = operating
+	data["inactive"] = FALSE
+	data["no_eject"] = FALSE
+	data["tooltip"] = ""
+	if(dirty >= 100)
+		data["inactive"] = TRUE
+		data["tooltip"] = "It's too dirty."
+	else if(!length(contents))
+		data["inactive"] = TRUE
+		data["no_eject"] = TRUE
+		data["tooltip"] = "Theres no contents."
+
+	var/list/items_counts = list()
+	var/list/name_overrides = list()
+	for(var/obj/O in contents)
+		var/display_name = O.name
+		if(istype(O,/obj/item/reagent_containers/food/snacks/egg)) // God I fuckin hate this
+			name_overrides[display_name] = "egg"
+			if(items_counts[display_name])
+				name_overrides[display_name] = "eggs"
+		if(istype(O,/obj/item/reagent_containers/food/snacks/tofu))
+			name_overrides[display_name] = "tofu chunk"
+			if(items_counts[display_name])
+				name_overrides[display_name] = "tofu chunks"
+		if(istype(O,/obj/item/reagent_containers/food/snacks/meat)) //any meat
+			name_overrides[display_name] = "slab of meat"
+			if(items_counts[display_name])
+				name_overrides[display_name] = "slabs of meat"
+		if(istype(O,/obj/item/reagent_containers/food/snacks/donkpocket))
+			display_name = "Turnovers"
+			name_overrides[display_name] = "turnover"
+			if(items_counts[display_name])
+				name_overrides[display_name] = "turnovers"
+		if(istype(O,/obj/item/reagent_containers/food/snacks/carpmeat))
+			name_overrides[display_name] = "fillet of meat"
+			if(items_counts[display_name])
+				name_overrides[display_name] = "fillets of meat"
+		items_counts[display_name]++
```

Comment:
All this stuff above is copied from the way stuff is examined, just some stuff had to be reformatted for info transfer to tgui. Any recommendations on how to improve this are greatly appreciated.